tests/widgetresize.js

```diff
@@ -0,0 +1,324 @@
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* globals document, Event */
+
+import WidgetResize from '../src/widgetresize';
+
+// ClassicTestEditor can't be used, as it doesn't handle the focus, which is needed to test resizer visual cues.
+import ClassicEditor from '@ckeditor/ckeditor5-editor-classic/src/classiceditor';
+import ArticlePluginSet from '@ckeditor/ckeditor5-core/tests/_utils/articlepluginset';
+
+import {
+	toWidget
+} from '../src/utils';
+import {
+	setData as setModelData
+} from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
+
+import Rect from '@ckeditor/ckeditor5-utils/src/dom/rect';
+
+describe( 'WidgetResize', () => {
+	let editor, editorElement, widget, mouseListenerSpies;
+
+	const commitStub = sinon.stub();
+	const mouseMock = {
+		down( editor, domTarget ) {
+			this._getPlugin( editor )._mouseDownListener( {}, {
+				target: domTarget
+			} );
+		},
+		move( editor, domTarget, eventData ) {
+			const combinedEventData = Object.assign( {}, eventData, {
+				target: domTarget
+			} );
+
+			this._getPlugin( editor )._mouseMoveListener( {}, combinedEventData );
+		},
+		up() {
+			this._getPlugin( editor )._mouseUpListener();
+		},
+
+		_getPlugin( editor ) {
+			return editor.plugins.get( WidgetResize );
+		}
+	};
+
+	before( () => {
+		mouseListenerSpies = {
+			down: sinon.spy( WidgetResize.prototype, '_mouseDownListener' ),
+			move: sinon.spy( WidgetResize.prototype, '_mouseMoveListener' ),
+			up: sinon.spy( WidgetResize.prototype, '_mouseUpListener' )
+		};
+	} );
+
+	after( () => {
+		for ( const stub of Object.values( mouseListenerSpies ) ) {
+			stub.restore();
+		}
+	} );
+
+	beforeEach( async () => {
+		editorElement = createEditorElement();
+		editor = await createEditor( editorElement );
+
+		setModelData( editor.model, '[<widget></widget>]' );
+
+		focusEditor( editor );
+
+		widget = editor.editing.view.document.getRoot().getChild( 0 );
+
+		for ( const stub of Object.values( mouseListenerSpies ) ) {
+			stub.resetHistory();
+		}
+		commitStub.resetHistory();
+
+		// It's crucial to have a precisely defined editor size for this test suite.
+		editor.editing.view.change( writer => {
+			const viewEditableRoot = editor.editing.view.document.getRoot();
+			writer.setAttribute( 'style', 'width: 400px; padding: 0px; overflow: hidden', viewEditableRoot );
+		} );
+	} );
+
+	afterEach( () => {
+		editorElement.remove();
+
+		if ( editor ) {
+			return editor.destroy();
+		}
+	} );
+
+	describe( 'plugin', () => {
+		it( 'is loaded', () => {
+			expect( editor.plugins.get( WidgetResize ) ).to.be.instanceOf( WidgetResize );
+		} );
+	} );
+
+	describe( 'mouse listeners', () => {
+		beforeEach( () => {
+			createResizer();
+		} );
+
+		it( 'doesnt break when called with unexpected element', async () => {
+			const unrelatedElement = document.createElement( 'div' );
+
+			editor.plugins.get( WidgetResize )._mouseDownListener( {}, {
```

Comment:
Here you test a single listener, not all of them (as the describe indicates).

Reply:
The "mouse listeners" test suite is meant to tests behavior of any mouse listeners. This can mean a single one or all of them. 

One could create a dedicated suite for mouse down listener alone, but it would be a separate group for only one test that has no significant value and is unlikely to get more unit tests in the future.

Reply:
The "mouse listeners" test suite is meant to tests behavior of any mouse listeners. This can mean a single one or all of them. 

One could create a dedicated suite for mouse down listener alone, but it would be a separate group for only one test that has no significant value and is unlikely to get more unit tests in the future.